src/expr.cpp

```diff
@@ -245,27 +309,71 @@ std::shared_ptr<Expr> ArithmeticExpr::convToBool(std::shared_ptr<Expr> arg) {
         arg, IntegralType::init(IntTypeID::BOOL), true);
 }
 
-std::shared_ptr<ArithmeticExpr>
-ArithmeticExpr::create(std::shared_ptr<PopulateCtx> ctx) {
+std::shared_ptr<Expr> ArithmeticExpr::create(std::shared_ptr<PopulateCtx> ctx) {
     auto gen_pol = ctx->getGenPolicy();
-    DataKind leaf_kind = rand_val_gen->getRandId(gen_pol->arith_leaf_distr);
-    std::shared_ptr<Expr> leaf;
-    if (leaf_kind == DataKind::VAR ||
-        ctx->getLocalSymTable()->getAvailSubs().empty()) {
-        size_t inp_var_idx = rand_val_gen->getRandValue(
-            static_cast<size_t>(0),
-            ctx->getExtInpSymTable()->getAvailVars().size() - 1);
-        leaf = ctx->getExtInpSymTable()->getAvailVars().at(inp_var_idx);
+    IRNodeKind node_kind = rand_val_gen->getRandId(gen_pol->arith_node_distr);
+    std::shared_ptr<Expr> new_node;
+    ctx->incArithDepth();
+    std::shared_ptr<PopulateCtx> active_ctx = ctx;
+    if (active_ctx->getArithDepth() == gen_pol->max_arith_depth) {
+        // We can heave only constants, variables and arrays as leaves
+        std::vector<Probability<IRNodeKind>> new_node_distr;
+        for (auto &item : gen_pol->arith_node_distr) {
+            if (item.getId() == IRNodeKind::CONST ||
+                item.getId() == IRNodeKind::SCALAR_VAR_USE ||
+                item.getId() == IRNodeKind::ARRAY_USE)
+                new_node_distr.push_back(item);
+        }
+
+        bool zero_prob = true;
+        for (auto &item : new_node_distr) {
+            if (item.getProb())
+                zero_prob = false;
+        }
+
+        if (zero_prob) {
+            for (auto &item : new_node_distr) {
+                item.increaseProb(30);
```

Comment:
For code like this, I'd like to see comments, which express your intent. Why increasing probability in case of zero-probabilility? Why increasing it by 30?

Reply:
I added comments and a better mechanism for heuristics like this.